File: test/e2e/Apps/BasicDotNetIsolated/PurgeOrchestrationHistory.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System.Net;
using Grpc.Core;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.DurableTask;
using Microsoft.DurableTask.Client;
using Microsoft.DurableTask.Entities;
using Microsoft.Extensions.Logging;

namespace Microsoft.Azure.Durable.Tests.E2E;

public static class PurgeOrchestrationHistory
{
    [Function(nameof(PurgeOrchestrationHistory))]
    public static async Task<HttpResponseData> PurgeHistory(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
        [DurableClient] DurableTaskClient client,
        FunctionContext executionContext,
        DateTime? purgeStartTime = null,
        DateTime? purgeEndTime = null,
        string? instanceId = null)
    {
        ILogger logger = executionContext.GetLogger("HelloCities_HttpStart");

        logger.LogInformation("Starting to purge instance histories");
        try
        {
            PurgeResult requestPurgeResult;
            if (!string.IsNullOrEmpty(instanceId))
            {
                // Purge a single instance
                requestPurgeResult = await client.PurgeInstanceAsync(instanceId);
                logger.LogInformation("Finished purging history for instance {instanceId}", instanceId);
            }
            else
            {
                // Purge by filter (terminal states only)
                requestPurgeResult = await client.PurgeAllInstancesAsync(new PurgeInstancesFilter(
                    purgeStartTime,
                    purgeEndTime,
                    [
                        OrchestrationRuntimeStatus.Completed,
                        OrchestrationRuntimeStatus.Failed,
                        OrchestrationRuntimeStatus.Terminated
                    ]));
                logger.LogInformation("Finished purge all instance history");
            }

            HttpResponseData response = req.CreateResponse(HttpStatusCode.OK);
            response.Headers.Add("Content-Type", "text/plain");
            await response.WriteStringAsync($"Purged {requestPurgeResult.PurgedInstanceCount} records");
            return response;
        }
        catch (InvalidOperationException ex)
        {
            HttpResponseData response = req.CreateResponse(HttpStatusCode.PreconditionFailed);
            response.Headers.Add("Content-Type", "text/plain");
            await response.WriteStringAsync(ex.Message);
            return response;
        }
        catch (RpcException ex)
        {
            logger.LogError(ex, "Failed to purge all instance history");
            HttpResponseData response = req.CreateResponse(HttpStatusCode.InternalServerError);
            response.Headers.Add("Content-Type", "text/plain");
            await response.WriteStringAsync($"Failed to purge all instance history: {ex.Message}");
            return response;
        }
    }

    [Function(nameof(InvokeDummyEntityOrchestration))]
    public static async Task<string> InvokeDummyEntityOrchestration([OrchestrationTrigger] TaskOrchestrationContext context)
    {
        var entityId = new EntityInstanceId(nameof(DummyEntity), "myEntity");
        await context.Entities.CallEntityAsync(entityId, string.Empty);
        return "Success";
    }

    [Function(nameof(DummyEntity))]
    public static Task DummyEntity([EntityTrigger] TaskEntityDispatcher dispatcher)
    {
        return dispatcher.DispatchAsync(operation =>
        {
            operation.State.SetState("state");
            return new(0);
        });
    }
}

File: test/e2e/Apps/BasicJava/src/main/java/com/function/PurgeOrchestrationHistory.java
package com.function;

import com.microsoft.azure.functions.annotation.*;
import com.microsoft.azure.functions.*;

import java.time.Instant;
import java.util.*;
import java.util.logging.Logger;

import com.microsoft.durabletask.*;
import com.microsoft.durabletask.azurefunctions.DurableClientContext;
import com.microsoft.durabletask.azurefunctions.DurableClientInput;


/**
 * Azure Functions with HTTP Trigger to purge orchestration history.
 */
public class PurgeOrchestrationHistory {
    /**
     * This HTTP-triggered function purges orchestration history.
     */
    @FunctionName("PurgeOrchestrationHistory")
    public HttpResponseMessage purgeHistory(
            @HttpTrigger(name = "req", methods = {HttpMethod.GET, HttpMethod.POST}, authLevel = AuthorizationLevel.ANONYMOUS) HttpRequestMessage<Optional<String>> request,
            @DurableClientInput(name = "durableContext") DurableClientContext durableContext,
            final ExecutionContext context) {

        Logger logger = context.getLogger();
        logger.info("Starting to purge instance histories");

        DurableTaskClient client = durableContext.getClient();

        // Attempt to parse the instance ID if just one instance is specified
        String instanceId = request.getQueryParameters().get("instanceId");
    
        try {
            PurgeResult purgeResult;
            if (instanceId != null && !instanceId.isEmpty())
            {
                purgeResult = client.purgeInstance(instanceId);
                logger.info("Finished purging history for instance " + instanceId);
            }
            else
            {
                // Parse optional query parameters for purgeStartTime and purgeEndTime
                String purgeStartTimeStr = request.getQueryParameters().get("purgeStartTime");
                String purgeEndTimeStr = request.getQueryParameters().get("purgeEndTime");
                Instant purgeStartTime = null;
                Instant purgeEndTime = null;
                try {
                    if (purgeStartTimeStr != null && !purgeStartTimeStr.isEmpty()) {
                        purgeStartTime = Instant.parse(purgeStartTimeStr);
                    }
                    if (purgeEndTimeStr != null && !purgeEndTimeStr.isEmpty()) {
                        purgeEndTime = Instant.parse(purgeEndTimeStr);
                    }
                } catch (Exception e) {
                    logger.severe("Invalid date format for purgeStartTime or purgeEndTime: " + e.getMessage());
                    return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                        .header("Content-Type", "text/plain")
                        .body("Invalid date format for purgeStartTime or purgeEndTime.")
                        .build();
                }

                PurgeInstanceCriteria filter = new PurgeInstanceCriteria();
                filter.setCreatedTimeFrom(purgeStartTime);
                filter.setCreatedTimeTo(purgeEndTime);
                filter.setRuntimeStatusList(Arrays.asList(
                    OrchestrationRuntimeStatus.COMPLETED,
                    OrchestrationRuntimeStatus.FAILED,
                    OrchestrationRuntimeStatus.TERMINATED
                ));

                purgeResult = client.purgeInstances(filter);

                logger.info("Finished purge all instance history");
            }

            return request.createResponseBuilder(HttpStatus.OK)
                .header("Content-Type", "text/plain")
                .body("Purged " + purgeResult.getDeletedInstanceCount() + " records")
                .build();
        } catch (Exception ex) {
            logger.severe("Failed to purge all instance history: " + ex.getMessage());
            return request.createResponseBuilder(HttpStatus.INTERNAL_SERVER_ERROR)
                .header("Content-Type", "text/plain")
                .body("Failed to purge all instance history: " + ex.getMessage())
                .build();
        }
    }
}